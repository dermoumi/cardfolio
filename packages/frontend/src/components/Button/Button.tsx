import type { FC, PropsWithChildren } from "react";

import styles from "./Button.module.css";

export type ButtonProps = PropsWithChildren<{
  onClick?: () => void;
  disabled?: boolean;
}>;

const Button: FC<ButtonProps> = ({ onClick, disabled, children }) => {
  return (
    <button className={styles.button} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};

export default Button;
