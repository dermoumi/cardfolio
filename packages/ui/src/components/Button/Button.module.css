.button {
  display: block;
  padding: 0.5rem 1rem;

  font-family: var(--font-family-sans);
  font-size: var(--font-size-md);
  font-weight: 500;

  user-select: none;

  border-radius: 999999px;
  border: 1px solid transparent;

  transition: all 0.1s ease-in-out;

  &:not([disabled]):active {
    transform: scale(0.98);
    transition: all 0.2s ease-in-out;
  }

  &[disabled] {
    border-style: dashed;
    border-color: var(--color-primary);
    background-color: var(--color-surface);
    color: var(--color-text);
  }
}

.primary {
  background-color: var(--color-primary);
  color: var(--color-primary-fg);

  &:not([disabled]) {
    &:hover {
      background-color: var(--color-primary-hover);
    }

    &:active {
      background-color: var(--color-primary-active);
    }
  }
}

.subtle {
  background-color: transparent;
  color: inherit;

  &:not([disabled]) {
    &:hover {
      background-color: var(--color-subtle-hover);
      color: var(--color-subtle-fg);
    }

    &:active {
      background-color: var(--color-subtle-active);
    }
  }
}
