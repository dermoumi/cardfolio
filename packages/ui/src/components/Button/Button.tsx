import type { FC } from "react";
import type { IconName } from "../Icon";

import classNames from "classnames";

import Icon from "../Icon";
import styles from "./Button.module.css";

const VARIANT_CLASSES = {
  primary: styles.primary,
  subtle: styles.subtle,
};

type CommonProps = {
  type?: "button" | "submit";
  onClick?: () => void;
  disabled?: boolean;
  variant?: keyof typeof VARIANT_CLASSES;
  icon?: IconName;
  label?: string;
};

export type ButtonProps =
  | CommonProps & { children: React.ReactNode; }
  | (CommonProps & { icon: IconName; label: string; children?: undefined; });

const Button: FC<ButtonProps> = (
  { type = "button", onClick, disabled, variant = "primary", children, icon, label },
) => {
  return (
    <button
      type={type}
      className={classNames(
        VARIANT_CLASSES[variant],
        styles.button,
        { [styles.iconOnly]: !children },
      )}
      onClick={onClick}
      disabled={disabled}
      aria-label={label}
    >
      {icon && <Icon name={icon} />}
      {children}
    </button>
  );
};

export default Button;
