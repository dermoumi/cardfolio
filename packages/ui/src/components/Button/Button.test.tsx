import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import Button from ".";
import styles from "./Button.module.css";

describe("Button", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<Button>Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with icon", () => {
    const { asFragment } = render(<Button icon="network">Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("has .icon-only class when no children", () => {
    const { getByRole } = render(<Button icon="network" label="Button" />);
    expect(getByRole("button").classList).toContain(styles.iconOnly);
  });

  it("calls onClick when clicked", () => {
    const onClick = vi.fn();
    const { getByText } = render(<Button onClick={onClick}>Button</Button>);

    getByText("Button").click();
    expect(onClick).toHaveBeenCalled();
  });
});
