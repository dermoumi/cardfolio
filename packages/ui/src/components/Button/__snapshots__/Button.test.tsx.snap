// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="_button_0829c1"
  >
    Button
  </button>
</DocumentFragment>
`;
